doc/BGP/Bgpcfgd-dyn-peer-modification-support.md

```diff
@@ -0,0 +1,144 @@
+
+
+# Definitions/Abbreviation
+###### Table 1: Abbreviations
+| Abbreviation             | Meaning                        |
+|--------------------------|--------------------------------|
+| BGP                      | Border Gateway Protocol        |
+| BGPCFGD                  | BGP configuration daemon       |
+| VNET                     | Virtual Network                |
+| VRF                      | Virtual Routing and Forwarding |
+
+
+# 1 Requirements Overview
+## 1.1 Functional requirements
+This section describes the SONiC requirements for supporting enhanced CRUD operations for dynamic bgp peers:
+- Configurable in VNET/VRFs, but equally applicable to default VRF
+- Dynamic peers' listen ranges must be modifiable at runtime such that ranges can be added and removed(changing today's behavior of being a create only attribute)
+- Dynamic peer updates/deletes should allow for changes to associated configurations like route-maps, peer-groups, prefix-lists etc
+- Ability to track the configuration of a dynamic BGP peer via a State DB entry which is queryable by the SDN controller
+- CLI commands to show VRF/VNET BGP neighbors
+
+## 1.2 Scalability requirements
+###### Table 2: Scalability
+| Component                | Expected value              |
+|--------------------------|-----------------------------|
+| Number of Dynamic BGP Peers each with route-map, prefix-list and peer-group|2k|
+| Number of listen ranges|4k|
+| Size of each Route-map, Prefix-list|<10|
+
+
+# 2 Modules Design
+
+## 2.1 Config DB
+Existing Config DB tables will be used to achieve this, no new Config DB table is introduced
+```
+BGP_PEER_RANGE|{{VRF/VNET-name}}|{{Peer-name}}:
+    "ip_range": [{{Range of IPs to add to listen range}}],
+    "name": {{Peer-name}},
+    "peer_asn": {{Peer's ASN number}}, (Optional)
+    "src_address": {{Src IP to initiate session}} (Optional)
+```
+
+## 2.2 State DB
+Following new table will be added to State DB.
+Use case: The SDN controller which programs neighbors will query the following state DB entry to confirm if the required entry is processed by bgpcfgd and programmed successfully. 
+```
+BGP_PEER_CONFIGURED_TABLE|{{VRF/VNET-name}}|{{Peer-name}}:
+    "ip_range": [{{Range of IPs to add to listen range}}],
+    "name": {{Peer-name}},
+    "peer_asn": {{Peer's ASN number}}, (Optional)
+    "src_address": {{Src IP to initiate session}} (Optional)
+```
+Unless otherwise stated, the attributes are mandatory.
+
+While the above shows the State DB schema for a dynamic peer, the schema for static peers bear the same table name format of ```BGP_PEER_CONFIGURED_TABLE|{{VRF/VNET-name}}|{{Peer-name}}```, but the key value pairs under the table will be those which are used to configure static peers.
+
+## 2.3 CLI
+The following CLIs will be added, each of these will have similar CLI outputs to their ```default VRF``` counterparts which are widely in use today:
+```
+1. show ip bgp vrf {vrf_name} summary
```

Comment:
Iso of adding new command can we add an option existing command `show ip bgp summary --vrf <vrf_name>` this way existing command `show ip bgp summary` will work for default VRF and can user can pass the vrf_name to get output per vrf.

Reply:
oh the existing command show ip bgp summary will still work for default vrf, intention is to have all existing commands work as-is, and support the extra vrf option in a way that it matches vtysh. Doing it this way will match exactly how the command is structured in vtysh, making it helpful for everyone who is used to that structure of cli.